extensions/typescript-language-features/src/tsServer/serverProcess.browser.ts

```diff
@@ -37,27 +37,72 @@ export class WorkerServerProcess implements TsServerProcess {
 	private readonly _onDataHandlers = new Set<(data: Proto.Response) => void>();
 	private readonly _onErrorHandlers = new Set<(err: Error) => void>();
 	private readonly _onExitHandlers = new Set<(code: number | null, signal: string | null) => void>();
+	/** For communicating with TS server synchronously */
+	private readonly tsserver: MessagePort;
+	/** For communicating watches asynchronously */
+	private readonly watcher: MessagePort;
+	/** For communicating with filesystem synchronously */
+	private readonly syncFs: MessagePort;
 
 	public constructor(
-		private readonly worker: Worker,
+		/** For logging and initial setup */
+		private readonly mainChannel: Worker,
+		extensionUri: vscode.Uri,
 		args: readonly string[],
 	) {
-		worker.addEventListener('message', (msg: any) => {
-			if (msg.data.type === 'log') {
-				this.output.append(msg.data.body);
+		const tsserverChannel = new MessageChannel();
+		const watcherChannel = new MessageChannel();
+		const syncChannel = new MessageChannel();
+		this.tsserver = tsserverChannel.port2;
+		this.watcher = watcherChannel.port2;
+		this.syncFs = syncChannel.port1;
+		this.tsserver.onmessage = (event) => {
+			if (event.data.type === 'log') {
+				console.error(`unexpected log message on tsserver channel: ${JSON.stringify(event)}`);
 				return;
 			}
-
 			for (const handler of this._onDataHandlers) {
-				handler(msg.data);
+				handler(event.data);
 			}
-		});
-		worker.onerror = (err: Error) => {
+		};
+		this.watcher.onmessage = (event) => {
+			console.error(`unexpected message on watcher channel: ${JSON.stringify(event)}`);
+		};
+		mainChannel.onmessage = (msg: any) => {
+			// for logging only
+			if (msg.data.type === 'log') {
+				this.output.append(msg.data.body);
+				return;
+			}
+			console.error(`unexpected message on main channel: ${JSON.stringify(msg)}`);
+			// TODO: A multi-watcher message would look something like this:
+			// if (msg.data.type === 'dispose-watcher') {
+			//	 watchers.get(msg.data.path).dispose()
+			// }
+		};
+		mainChannel.onerror = (err: ErrorEvent) => {
+			this.output.append(JSON.stringify('error! ' + JSON.stringify(err)) + '\n');
 			for (const handler of this._onErrorHandlers) {
-				handler(err);
+				// TODO: The ErrorEvent type might be wrong; previously this was typed as Error and didn't have the property access.
+				handler(err.error);
 			}
 		};
-		worker.postMessage(args);
+		// TODO: For prototyping, create one watcher ahead of time and send messages using the worker.
+		// For the real thing, it makes more sense to create a third MessageChannel and listen on it; the host
+		// can then send messages to tell the extension to create a new filesystemwatcher for each file/directory
+		const fsWatcher = vscode.workspace.createFileSystemWatcher('**/*');
```

Comment:
We may be able to reuse the markdown watcher manager for this: https://github.com/microsoft/vscode/blob/ab1a31add4608f484d555e3c791f03321a4dda5f/extensions/markdown-language-features/src/client/fileWatchingManager.ts#L18